brainiak/fcma/classifier.py

```diff
@@ -199,82 +304,54 @@ def predict(self, X):
-            data = np.zeros((num_test_samples, num_training_samples),
-                            np.float32,
-                            order='C')
-            corr_data = corr_data.reshape(num_test_samples,
-                                          num_voxels * num_voxels)
-            # compute the similarity matrix using corr_data and training_data
-            blas.compute_single_matrix_multiplication('T', 'N',
-                                                      num_training_samples,
-                                                      num_test_samples,
-                                                      num_voxels * num_voxels,
-                                                      1.0,
-                                                      self.training_data_,
-                                                      num_voxels * num_voxels,
-                                                      corr_data,
-                                                      num_voxels * num_voxels,
-                                                      0.0,
-                                                      data,
-                                                      num_training_samples)
-            logger.debug(
-                'similarity matrix computation done'
-            )
-        else:
-            data = corr_data.reshape(num_test_samples,
-                                     num_voxels * num_voxels)
-        y_pred = self.clf.predict(data)
+        # correlation computation
+        corr_data = self._prepare_corerelation_data(X)
+        # normalization
+        corr_data = self._normalize_correlation_data(corr_data, len(X))
+        # test data generation
+        self.test_data_ = self._prepare_test_data(corr_data)
+        # prediction
+        y_pred = self.clf.predict(self.test_data_)
         time2 = time.time()
         logger.info(
             'prediction done, takes %.2f s' %
             (time2 - time1)
         )
         return y_pred
+
+    def decision_function(self, X, has_test_data=False):
+        """ output the decision value of the prediction
+
+        if the test_data is not computed, i.e. predict is not called,
+        first generate the test_data
+        after getting the test_data, get the decision value via self.clf.
+
+        Parameters
+        ----------
+        X: a list of numpy array in shape [num_TRs, self.num_voxels\_]
+            X contains the activity data filtered by top voxels
+            and prepared for correlation computation.
+            len(X) is the number of test samples
+            if len(X) > 1: normalization is done
+            on all test samples
+        has_test_data: Boolean value, default False
+            specify if the test_data has been generated
+
+        Returns
+        -------
+        confidence: the predictions confidence values of X, in shape [len(X),]
+        """
+        if has_test_data:
+            assert len(X) == len(self.test_data_), \
```

Comment:
I understand what `X` is, that is why I proposed calling storing it as `self.raw_test_data_`.

In the current proposal, if the user calls `decision_function(X1, False)` and then mistakenly calls `decision_function(X2, True)` (instead of `decision_function(X2, False)`), the result will be wrong if `len(X1) == len(X2)`.

With my proposal, the interface will be simpler (no more `has_test_data` parameter) and the check will ensure old test data is not reused.